File: solutions/erlang/bob/2/src/bob.erl
-module(bob).

-export([response/1]).

response(String) ->
    Trimmed = string:trim(String),
    Upcase = string:uppercase(String),
    Locase = string:lowercase(String),
    Yelled = (Upcase == String) and (Locase /= String),
    Question =
        length(Trimmed) andalso string:slice(Trimmed, string:length(Trimmed) - 1, 1) == "?",
    response(Trimmed, Yelled, Question).

response(_Trimmed, true, true) ->
    "Calm down, I know what I'm doing!";
response(_Trimmed, true, _Question) ->
    "Whoa, chill out!";
response(_Trimmed, _Yelled, true) ->
    "Sure.";
response("", _Yelled, _Question) ->
    "Fine. Be that way!";
response(_Trimmed, _Yelled, _Question) ->
    "Whatever.".

File: solutions/elixir/bob/2/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      Regex.run(~r/^[\p{Lu}' ]*\?[ ]*$/u, input) != nil ->
        # yelled questions
        "Calm down, I know what I'm doing!"

      Regex.run(~r/^(?=.*\p{Lu}{1,})[\p{Lu}0-9![:punct:] ]*$/u, input) != nil ->
        # shouting
        "Whoa, chill out!"

      Regex.run(~r/^[\p{L}[:punct:]0-9 ]*\?[ ]*$/u, input) != nil ->
        # questions
        "Sure."

      Regex.run(~r/^[\n\r \t]*$/u, input) != nil ->
        # Exclamation
        "Fine. Be that way!"

      true ->
        "Whatever."
    end
  end
end
